lib/services/cloud-codesign-service.ts

```diff
@@ -0,0 +1,133 @@
+		super($fs, $httpClient, $logger);
+		this.failedToStartError = "Failed to start cloud codesign generation.";
+		this.failedError = "Codesign generation failed.";
+		this.operationInProgressStatus = "Working";
+	}
+
+	public async generateCodesignFiles(codesignData: ICodesignData,
+		projectData: IProjectData): Promise<ICodesignResultData> {
+		codesignData.clean = codesignData.clean === undefined ? true : codesignData.clean;
+		const buildId = uuid.v4();
+
+		try {
+			const serverResult = await this.executeGeneration(codesignData, projectData, buildId);
+			return serverResult;
+		} catch (err) {
+			err.buildId = buildId;
+			throw err;
+		}
+	}
+
+	public getServerOperationOutputDirectory(options: ICloudServerOutputDirectoryOptions): string {
+		return path.join(options.projectDir, constants.CODESIGN_TEMP_DIR_NAME, options.platform.toLowerCase());
+	}
+
+	protected getServerResults(codesignResult: IServerResult): IServerItem[] {
+		const result = _.filter(codesignResult.buildItems, b => b.disposition === constants.DISPOSITIONS.CERTIFICATE
+			|| b.disposition === constants.DISPOSITIONS.PROVISION);
+
+		if (!result) {
+			this.$errors.failWithoutHelp(
+				`No item with disposition ${constants.DISPOSITIONS.CERTIFICATE} or ${constants.DISPOSITIONS.PROVISION} found in the server result items.`);
+		}
+
+		return result;
+	}
+
+	protected async getServerLogs(logsUrl: string, buildId: string): Promise<void> {
+		// no specific implementation needed.
+	}
+
+	public async executeGeneration(codesignData: ICodesignData,
+		projectData: IProjectData,
+		buildId: string): Promise<ICodesignResultData> {
+		const codesignInformationString = "Generation of iOS certificate and provision files";
+		this.$logger.info(`Starting ${codesignInformationString}.`);
+
+		const codesignRequest = await this.prepareCodesignRequest(buildId, codesignData, projectData);
+		const codesignResponse: IServerResponse = await this.$buildCloudService.generateCodesignFiles(codesignRequest);
+		this.$logger.trace(`Codesign response: ${JSON.stringify(codesignResponse)}`);
+
+		let codesignResult: IServerResult;
+		try {
+			await this.waitForServerOperationToFinish(buildId, codesignResponse);
+		} catch (ex) {
+			this.$logger.trace("Codesign generation failed with err: ", ex);
+		} finally {
+			codesignResult = await this.getObjectFromS3File<IServerResult>(codesignResponse.resultUrl);
+		}
+
+		this.$logger.trace("Codesign result:");
+		this.$logger.trace(codesignResult);
+
+		if (!codesignResult.buildItems || !codesignResult.buildItems.length) {
+			// Something failed.
+			this.$errors.failWithoutHelp(`Codesign failed. Reason is: ${codesignResult.errors}. Additional information: ${codesignResult.stderr}.`);
+		}
+
+		this.$logger.info(`Finished ${codesignInformationString} successfully. Downloading result...`);
+
+		const localCodesignResults = await this.downloadServerResults(codesignResult, {
+			projectDir: projectData.projectDir,
+			platform: this.platform,
```

Comment:
In case we'll use the same class for Android generation in the future, you should not use the platform from `this`, but receive it from the caller.

Reply:
Of course. It will not be hard-coded as well. Currently I added it for consistency - having the folders structure with iOS in it is neat for iOS codesign generation. I just laid some foundations in case we'll have other platforms. Further refactoring will be needed at such case.